sklearn/manifold/_barnes_hut_tsne.pyx

```diff
@@ -0,0 +1,682 @@
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: cdivision=True
+# Author: Christopher Moody <chrisemoody@gmail.com>
+# Author: Nick Travers <nickt@squareup.com>
+# Implementation by Chris Moody & Nick Travers
+# See http://homepage.tudelft.nl/19j49/t-SNE.html for reference
+# implementations and papers describing the technique
+
+
+from libc.stdlib cimport malloc, free
+from libc.stdio cimport printf
+from libc.math cimport sqrt
+cimport numpy as np
+import numpy as np
+
+cdef extern from "math.h":
+    float fabsf(float x) nogil
+
+# Round points differing by less than this amount
+# effectively ignoring differences near the 32bit 
+# floating point precision
+cdef float EPSILON = 1e-6
+
+# This is effectively an ifdef statement in Cython
+# It allows us to write printf debugging lines
+# and remove them at compile time
+cdef enum:
+    DEBUGFLAG = 0
+
+cdef extern from "time.h":
+    # Declare only what is necessary from `tm` structure.
+    ctypedef long clock_t
+    clock_t clock() nogil
+    double CLOCKS_PER_SEC
+
+
+cdef struct Node:
+    # Keep track of the center of mass
+    float[3] cum_com
```

Comment:
what does `cum` stand for? "cumulative"? I think this field could be renamed `center_of_mass` or `cumulative_center_of_mass` if there is an ambiguity. Or just `center` if space is an issue (but it does not seem to be the case.